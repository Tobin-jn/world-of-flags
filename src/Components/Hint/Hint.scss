.Hint {
  background: rgba(240, 240, 240, .99);
  border-radius: 8px;
  box-shadow: 0 0 1rem rgba(0, 0, 0, .3);
  display: inline-flex;
  flex-direction: column;
  height: 55%;
  justify-content: space-between;
  left: 2%;
  padding: 5rem 0;
  position: absolute;
  text-align: center;
  top: 8rem;
  width: 96%;
  z-index: 2;
}

p {
  font-size: 2rem;
  margin: 0;
}

.hint-title {
  font-size: 3rem;
}