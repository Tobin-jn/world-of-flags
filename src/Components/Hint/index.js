import React, { Component } from 'react';
import { Route, NavLink } from 'react-router-dom';
import './Hint.scss';

class Hint extends Component {
  constructor(props) {
    super(props);
  }

  handleClick = () => {
    this.props.hideHint()
  }

  getCountryOutlinePath = () => {
    const { outline } = this.props;
    const outlineUrl = `https://flagz4u.herokuapp.com${outline}`
    return outlineUrl
  }

  showHint = () => {
    const { hint, outline, fact } = this.props;

    if(hint === 'fact') {
      return <p className='hint-fact'>{fact}</p>
    } else if (hint === 'outline') {
        const outline = this.getCountryOutlinePath()
        return <img src={outline} className='hint-outline'/>
    } else if(hint === 'out of hints') {
        return <p className='hint-fact out-of-hints'>You are out of hints!</p>
    }
  }
  render() {
    return(
      <div className='Hint'>
        <p className='hint-title'>Hint</p>
        {this.showHint()}
        <div 
          className='button hint-continue' 
          onClick={this.handleClick}>Continue
        </div>
      </div>
    )
  }
}

Hint.propTypes = {
  hint: PropTypes.string.isRequired,
  outline: PropTypes.string.isRequired,
  fact: PropTypes.string.isRequired,
  hideHint: PropTypes.func.isRequired,
}

export default Hint;