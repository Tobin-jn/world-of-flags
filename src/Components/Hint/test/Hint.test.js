import React from 'react'
import { shallow, mount } from 'enzyme'
import Hint from '../index'

describe('Hint', () => {
  let wrapper
  let mockCurrentCountry
  let mockHint

  beforeEach(() => {
    mockCurrentCountry = {
      id: 46,
      name: 'The Netherlands',
      flag: '/images/flags/netherlands.png',
      country_outline: '/images/outlines/netherlands.png',
      created_at: '2018-12-28T12:02:21.458Z',
      updated_at: '2018-12-28T12:02:21.458Z',
      multipleChoice: ['Seychelles', 'Malawi', 'Cambodia', 'The Netherlands'],
      facts: {
        id: 106,
        country_fact:
          'Numerous dikes cover the coast of Ijsselmeer in this country',
        country_id: 46,
        created_at: '2018-12-28T12:02:21.951Z',
        updated_at: '2018-12-28T12:02:21.951Z',
      },
    }
    mockHint = 'fact'
    wrapper = shallow(
      <Hint
        hideHint={jest.fn()}
        hint={mockHint}
        outline={mockCurrentCountry.country_outline}
        fact={mockCurrentCountry.facts.country_fact}
      />
    )
  })

  it('should render like the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('handleClick', () => {
    it('should invoke hideHint', () => {
      const props = wrapper.instance().props
      wrapper.instance().handleClick()
      const spy = jest.spyOn(props, 'hideHint')
      expect(spy).toHaveBeenCalled()
    })
  })

  describe('getCountryOutlinePath', () => {
    it('should create complete path for country outline', () => {
      const expected =
        'https://flagz4u.herokuapp.com/images/outlines/netherlands.png'
      const result = wrapper.instance().getCountryOutlinePath()
      expect(result).toEqual(expected)
    })
  })

  describe('showHint', () => {
    it('should show a country fact as a first hint', () => {
      expect(wrapper.exists('.hint-fact')).toEqual(true)
    })

    it('should invoke getCountryOutlinePath as a second hint', () => {
      mockHint = 'outline'
      wrapper = shallow(
        <Hint
          hideHint={jest.fn()}
          hint={mockHint}
          outline={mockCurrentCountry.country_outline}
          fact={mockCurrentCountry.facts.country_fact}
        />
      )
      expect(wrapper.exists('.hint-outline')).toEqual(true)
    })

    it('should notify user that they are out of hints after 2nd hint', () => {
      mockHint = 'out of hints'
      wrapper = shallow(
        <Hint
          hideHint={jest.fn()}
          hint={mockHint}
          outline={mockCurrentCountry.country_outline}
          fact={mockCurrentCountry.facts.country_fact}
        />
      )
      expect(wrapper.exists('.out-of-hints')).toEqual(true)
    })
  })
})
