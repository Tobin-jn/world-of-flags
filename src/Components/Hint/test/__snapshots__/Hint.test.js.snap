// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hint should render like the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Hint />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="hint-title"
        >
          Hint
        </p>,
        <div
          className="button hint-continue"
          onClick={[Function]}
        >
          Continue
        </div>,
      ],
      "className": "Hint",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Hint",
          "className": "hint-title",
        },
        "ref": null,
        "rendered": "Hint",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Continue",
          "className": "button hint-continue",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Continue",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="hint-title"
          >
            Hint
          </p>,
          <div
            className="button hint-continue"
            onClick={[Function]}
          >
            Continue
          </div>,
        ],
        "className": "Hint",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Hint",
            "className": "hint-title",
          },
          "ref": null,
          "rendered": "Hint",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Continue",
            "className": "button hint-continue",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Continue",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
